src/template.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  ARROW_CLASS,
+  BACKDROP_CLASS,
+  BOX_CLASS,
+  CONTENT_CLASS,
+  SVG_ARROW_CLASS,
+} from './constants';
+import {div, isElement} from './dom-utils';
+import {Instance, PopperChildren, PopperElement, Props} from './types';
+import {arrayFrom} from './utils';
+
+const innerHTML = (): 'innerHTML' => 'innerHTML';
+
+function dangerouslySetInnerHTML(element: Element, html: string): void {
+  element[innerHTML()] = html;
+}
+
+function createArrowElement(value: Props['arrow']): HTMLDivElement {
+  const arrow = div();
+
+  if (value === true) {
+    arrow.className = ARROW_CLASS;
+  } else {
+    arrow.className = SVG_ARROW_CLASS;
+
+    if (isElement(value)) {
+      arrow.appendChild(value);
+    } else {
+      dangerouslySetInnerHTML(arrow, value as string);
+    }
+  }
+
+  return arrow;
+}
+
+export function setContent(
+  content: HTMLDivElement,
+  {props, reference}: Instance,
+): void {
+  const computedContent =
+    typeof props.content === 'function'
+      ? props.content(reference)
+      : props.content;
+
+  if (isElement(computedContent)) {
+    dangerouslySetInnerHTML(content, '');
+    content.appendChild(computedContent);
+  } else {
+    if (props.allowHTML) {
+      dangerouslySetInnerHTML(content, computedContent);
+    } else {
+      content.textContent = computedContent;
+    }
+  }
+}
+
+export function getChildren(popper: PopperElement): PopperChildren {
+  const box = popper.firstElementChild as HTMLDivElement;
+  const boxChildren = arrayFrom(box.children);
+
+  return {
+    box,
+    content: boxChildren.find(node => node.classList.contains(CONTENT_CLASS)),
+    arrow:
+      boxChildren.find(
+        node =>
+          node.classList.contains(ARROW_CLASS) ||
+          node.classList.contains(SVG_ARROW_CLASS),
+      ) || null,
+    backdrop:
+      boxChildren.find(node => node.classList.contains(BACKDROP_CLASS)) || null,
```

Comment:
Does it have to default to `null`? I think `undefined` would be just fine if not found.